### PR TITLE
Rewrite Models and subsequently Routes in Typescript

### LINKED ISSUE #1757: [🧹] Rewrite sequelize models in proper TypeScript
A significant amount of remaining TypeScript errors comes from the way the Sequelize models are currently defined and created. The models defined in `/models` as well as the dynamic creation in `/models/index.ts` should be rewritten in TypeScript-style.

## Recommended steps

1. Inspect files in https://github.com/juice-shop/juice-shop/tree/master/models to understand the current implementation
2. Run `npm run build:server` and analyze TypeScript errors for function calls on `models.Challenges`, `models.User` and other models
2. Read https://sequelize.org/master/manual/typescript.html to understand implementation recommended by Seqeuelize
3. Identify necessary migration steps
4. Gradually migrate all models and adapt all places where they are used to the new implementation

### PR DESCRIPTION
**Pull request by ShubhamPalriwala**
Closes #1757 

#### Tasks Done:
- [x] Rewrite Models in Typescript
    - Sequelize currently does not currently handle foreign keys in the best way possible in Typescript
    - I discussed this with the community and maintainers on their Slack and have used the best possible solution as of now according to our discussion [here](https://sequelize.slack.com/archives/C0CELV3N1/p1644759779263599), can be accessed once you join their workspace using their inviter bot [here](https://github.com/sequelize/sequelize#book-resources).
    - An issue has also been opened based on our discussion  https://github.com/sequelize/sequelize/issues/14125
- [x] Rewrite controllers for the routes in Typescript
    - Made some minor typing tweaks and they were fixed
    - One specific unique handling case was required in the `routes/payment.ts` where we had to mask the `cardNum` property using `*` hence had to be converted to string. To solve this, a new interface `displayCard` has been created just for this specific use case.

Rebased my branch just before opening the Pull Request to make sure it's up-to-date.

#### The commits are as follows:
- [Commit 1](295f89b1105b28a3b013667e8dd98fb2939cd57f): Convert Models to class-based typescript
- [Commit 2](295f89b1105b28a3b013667e8dd98fb2939cd57f): Fix the typings in `datacreator.ts` and update models wherever required
- [Commit 3](19edb5ee6a782d944a6c626996528f2db1901ec7): Fix the typings in the `routes/` directory related to the models
- [Commit 4](6a1936289c196b6815bb8c37c492a26817725b01): Cleanup unused imports and commented out code from `models/`
- [Commit 5](299ec4252d46a06682a1923dd3693123dd184366): Used the `npm run lint:fix` to lint the files and one type conversion from string to number

Edit: Commits continued while fixing the CI bugs hence lost track. The above are invalid now, but the PR is finally valid including the CI checks :))

### CODE REVIEW

**Comment by bkimminich** [comment]
Just approved CI/CD, so let's 🚀!

**Comment by ShubhamPalriwala** [comment]
Looks like there are a number of CI/CD processes failing. Sorry for this, will update soon and ping you here once it's ready to be reviewed!

**Comment by ShubhamPalriwala** [comment]
Update: Have enabled workflows on my fork and fixed all the linting errors.

However, the other ones still fail. Gonna take some rest now and start tomorrow with the `coding-challenge-rsn`.

I did a little bit of debugging and have a query here:
Am I also supposed to **manually change the codebase in the `data/static/` directory**? Because as of now I think that is the only thing that is importing the models the old way hence why the test cases might be failing!

Edit: Some files in `test/e2e/` directory do too!


PS: Please do not approve the workflows in this PR yet because it's anyways gonna fail and cost us Github Action minutes so I'll let y'all know when to!

**Comment by bkimminich** [comment]
You can ignore the `coding-challenges-rsn`, I'm happy to fix those after the merge. Anything in the `data/static/codefixes` folder should be excluded from TypeScript, so it shouldn't cause any CI/CD issues other than the mentioned `coding-challenges-rsn`. See https://pwning.owasp-juice.shop/appendix/code-snippets.html if you're interested, but again - I'm happy to fix those issues later.

**Comment by bkimminich** [comment]
PS: Action minutes should not be an issue. https://github.com/juice-shop thankfully has a free _Team_ plan sponsored from GitHub with seemingly endless minutes, as we spent 0 or 3000min in the ongoing billing period.

**Comment by ShubhamPalriwala** [comment]
Update for today, fixed the  failing `test` Github Action but the e2e is giving me a good amount of trouble. Gotta prepare for my exam tomorrow hence will continue tomorrow only.

Sorry for the delay due to this!

**Comment by bkimminich** [comment]
I don't see any `e2e` CI runs on your fork and the last one on the PR failed from a model import. If any actual e2e tests fail, it means the rewrite impacted hacking challenges. I could imagine this happening in SQLi related challenges if the newer Sequielize version has a stricter default security model, but other challenges should - in theory - not be affected. I'll gladly hekp you look into `e2e` fails once I see one here on your PR.

**Comment by ShubhamPalriwala** [comment]
Yeah! I fixed the imports hence the `test` Action started working.
The sequelize upgrade might've caused this. will take a look, thank you.

I'm running the e2e tests locally using `npm run e2e` hence I didn't push it.
Give me a couple of days to try things out and then I will reach out to you with whatever I still face issues in. 🙏🏼 

Thank you for helping me out might ping you soon!

**Comment by bkimminich** [comment]
I've pushed the linted/styled code back to the PR, because working on diffs where one side uses a completely different codestyle is not feasible.

File: models/recycle.ts [hunk 1]
```diff
@@ -4,17 +4,52 @@
...
+      AddressId: {
+        type: DataTypes.INTEGER
+      },
+      id: {
+        type: DataTypes.INTEGER,
+        primaryKey: true,
+        autoIncrement: true
+      },
+      // TODO set this as DataTypes.INTEGER(4)
+      quantity: DataTypes.INTEGER,
```
Comment (ShubhamPalriwala) [inline models/recycle.ts:43]: @bkimminich We previusly had this as 
```
    quantity: INTEGER(4),
```
but this does not work in Typescript. However, all the test cases pass with the current implementation in the PR.

Is there any specific reason why this was set as 4?

Comment (ShubhamPalriwala) [reply models/recycle.ts:43]: As discussed on slack, this limitation is not enforced anywhere hence we go ahead without the 4 limit.

**Pull request merged**